File: backend/app/schemas/risk.py
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class RiskPolicyWriteRequest(BaseModel):
    staking_mode: Literal["flat_percent", "fractional_kelly"]
    flat_stake_pct: Decimal | None = Field(default=None, gt=0, le=Decimal("0.05"))
    kelly_fraction: Decimal | None = Field(default=None, gt=0, le=Decimal("0.50"))
    max_ticket_pct: Decimal = Field(gt=0, le=Decimal("0.05"))
    max_open_exposure_pct: Decimal = Field(gt=0, le=Decimal("0.20"))
    max_match_pct: Decimal = Field(gt=0, le=Decimal("0.20"))
    max_team_pct: Decimal = Field(gt=0, le=Decimal("0.20"))
    max_league_window_pct: Decimal = Field(gt=0, le=Decimal("0.20"))
    league_window_hours: int = Field(gt=0, le=24)
    max_daily_stake_pct: Decimal = Field(gt=0, le=Decimal("1.00"))
    max_weekly_stake_pct: Decimal = Field(gt=0, le=Decimal("1.00"))
    max_daily_ticket_count: int = Field(gt=0, le=1000)
    max_weekly_ticket_count: int = Field(gt=0, le=5000)
    accumulators_enabled: bool
    automation_enabled: bool

    @model_validator(mode="after")
    def validate_complete_policy(self):
        if self.staking_mode == "flat_percent":
            if self.flat_stake_pct is None or self.kelly_fraction is not None:
                raise ValueError("flat_percent requires flat_stake_pct and forbids kelly_fraction")
        elif self.kelly_fraction is None or self.flat_stake_pct is not None:
            raise ValueError("fractional_kelly requires kelly_fraction and forbids flat_stake_pct")
        if self.max_daily_stake_pct > self.max_weekly_stake_pct:
            raise ValueError("max_daily_stake_pct cannot exceed max_weekly_stake_pct")
        if self.max_daily_ticket_count > self.max_weekly_ticket_count:
            raise ValueError("max_daily_ticket_count cannot exceed max_weekly_ticket_count")
        for value, field in (
            (self.max_match_pct, "max_match_pct"),
            (self.max_team_pct, "max_team_pct"),
            (self.max_league_window_pct, "max_league_window_pct"),
        ):
            if value > self.max_open_exposure_pct:
                raise ValueError(f"{field} cannot exceed max_open_exposure_pct")
        if self.flat_stake_pct is not None and self.flat_stake_pct > self.max_ticket_pct:
            raise ValueError("flat_stake_pct cannot exceed max_ticket_pct")
        return self


class RiskPolicyResponse(RiskPolicyWriteRequest):
    id: int
    bankroll_id: int
    version: int
    effective_from: datetime
    created_at: datetime


class RiskUsageResponse(BaseModel):
    bankroll_balance: Decimal | None = None
    available_balance: Decimal | None = None
    open_exposure_amount: Decimal = Decimal("0.00")
    open_exposure_pct: Decimal | None = None
    staked_last_24h: Decimal = Decimal("0.00")
    staked_last_7d: Decimal = Decimal("0.00")
    ticket_count_last_24h: int = 0
    ticket_count_last_7d: int = 0


class RiskPauseStateResponse(BaseModel):
    paused_until: datetime | None = None
    pause_reason: str | None = None
    updated_at: datetime | None = None


class RiskPolicyEnvelope(BaseModel):
    policy: RiskPolicyResponse | None = None
    pending_policy: RiskPolicyWriteRequest | None = None
    pending_effective_at: datetime | None = None
    state: RiskPauseStateResponse | None = None
    usage: RiskUsageResponse = Field(default_factory=RiskUsageResponse)
    hard_max_ticket_pct: Decimal = Decimal("0.05")
    hard_max_open_exposure_pct: Decimal = Decimal("0.20")


class RiskPauseRequest(BaseModel):
    paused_until: datetime
    pause_reason: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def validate_timezone(self):
        if self.paused_until.tzinfo is None or self.paused_until.utcoffset() is None:
            raise ValueError("paused_until must be timezone-aware")
        return self

File: frontend/src/lib/api/risk.ts
import { ApiClient, ApiClientError } from './client';

export type RiskStakingMode = 'flat_percent' | 'fractional_kelly';

export type RiskPolicy = {
	id: number;
	bankroll_id: number;
	version: number;
	staking_mode: RiskStakingMode;
	flat_stake_pct: number | null;
	kelly_fraction: number | null;
	max_ticket_pct: number;
	max_open_exposure_pct: number;
	max_match_pct: number | null;
	max_team_pct: number | null;
	max_league_window_pct: number | null;
	league_window_hours: number;
	max_daily_stake_pct: number | null;
	max_weekly_stake_pct: number | null;
	max_daily_ticket_count: number | null;
	max_weekly_ticket_count: number | null;
	accumulators_enabled: boolean;
	automation_enabled: boolean;
	effective_from: string;
	superseded_at?: string | null;
	created_at?: string;
};

export type RiskPolicyInput = {
	staking_mode: RiskStakingMode;
	flat_stake_pct: number | null;
	kelly_fraction: number | null;
	max_ticket_pct: number;
	max_open_exposure_pct: number;
	max_match_pct: number;
	max_team_pct: number;
	max_league_window_pct: number;
	league_window_hours: number;
	max_daily_stake_pct: number;
	max_weekly_stake_pct: number;
	max_daily_ticket_count: number;
	max_weekly_ticket_count: number;
	accumulators_enabled: boolean;
	automation_enabled: boolean;
};

export type RiskUsage = {
	bankroll_balance?: number | null;
	available_balance?: number | null;
	open_exposure_amount?: number | null;
	open_exposure_pct?: number | null;
	staked_last_24h?: number | null;
	staked_last_7d?: number | null;
	ticket_count_last_24h?: number | null;
	ticket_count_last_7d?: number | null;
};

export type RiskPauseState = {
	paused_until: string | null;
	pause_reason: string | null;
	updated_at?: string | null;
};

export type RiskPolicyOverview = {
	policy: RiskPolicy | null;
	pending_policy: RiskPolicy | null;
	state: RiskPauseState | null;
	usage: RiskUsage | null;
};

export type PauseBankrollInput = {
	paused_until: string;
	pause_reason: string;
};

type RiskPolicyApiResponse =
	| RiskPolicy
	| {
			policy?: RiskPolicy | null;
			active_policy?: RiskPolicy | null;
			pending_policy?: RiskPolicy | null;
			state?: RiskPauseState | null;
			pause?: RiskPauseState | null;
			usage?: RiskUsage | null;
	  };

function hasPolicyShape(value: RiskPolicyApiResponse): value is RiskPolicy {
	return 'staking_mode' in value && 'max_ticket_pct' in value;
}

export function normalizeRiskPolicyOverview(
	response: RiskPolicyApiResponse | null
): RiskPolicyOverview | null {
	if (response === null) return null;
	if (hasPolicyShape(response)) {
		return { policy: response, pending_policy: null, state: null, usage: null };
	}

	return {
		policy: response.policy ?? response.active_policy ?? null,
		pending_policy: response.pending_policy ?? null,
		state: response.state ?? response.pause ?? null,
		usage: response.usage ?? null
	};
}

function bankrollPath(bankrollId: number, suffix: string): string {
	if (!Number.isInteger(bankrollId) || bankrollId <= 0) {
		throw new TypeError('bankrollId must be a positive integer');
	}
	return `/api/v1/bankroll/${bankrollId}/${suffix}`;
}

export class RiskApi extends ApiClient {
	async getRiskPolicy(bankrollId: number): Promise<RiskPolicyOverview | null> {
		try {
			const response = await this.get<RiskPolicyApiResponse>(bankrollPath(bankrollId, 'risk-policy'));
			return normalizeRiskPolicyOverview(response);
		} catch (error) {
			if (error instanceof ApiClientError && error.statusCode === 404) return null;
			throw error;
		}
	}

	async saveRiskPolicy(bankrollId: number, input: RiskPolicyInput): Promise<RiskPolicyOverview> {
		const response = await this.put<RiskPolicyApiResponse>(
			bankrollPath(bankrollId, 'risk-policy'),
			input as unknown as Record<string, unknown>
		);
		return normalizeRiskPolicyOverview(response) ?? {
			policy: null,
			pending_policy: null,
			state: null,
			usage: null
		};
	}

	async pauseBankroll(bankrollId: number, input: PauseBankrollInput): Promise<RiskPolicyOverview> {
		const response = await this.post<RiskPolicyApiResponse>(
			bankrollPath(bankrollId, 'pause'),
			input as unknown as Record<string, unknown>
		);
		return normalizeRiskPolicyOverview(response) ?? {
			policy: null,
			pending_policy: null,
			state: null,
			usage: null
		};
	}
}

export const riskApi = new RiskApi();
